### DISCUSSION TITLE
[RFC][Runtime] Runtime Location API

### DISCUSSION THREAD

**Discussion by EvanBacon** [expo/expo#14385] [Proposals and Ideas]
# Issues

> [Stable whatwg spec](https://html.spec.whatwg.org/multipage/history.html#location)

Bundlers and other JavaScript tools/packages rely heavily on `window.location` API being available, in the native runtime (added by the browser) and there is seemingly no reason we can't add it to the Expo platform via a JSI module. The biggest question is how useful would this be, and what kinds of overhead would this add.

Almost all JS Runtimes support window.location, including Deno and Electron ([compatibility](https://developer.mozilla.org/en-US/docs/Web/API/Location#browser_compatibility)).

# Proposal

The JS bundle always comes from somewhere (dev server, remote hosting service, embedded in the binary), we should have an API to make the JS aware of where it hails from. By closely recreating the `Location` spec, we can ensure a stable API with backwards compatibility that developers can safely rely on when using the Expo platform for their apps.

## Properties

In Electron, `location.href` refers to the development server URL `http://localhost:19000/`, and a local file path in production like `file://Users/path/to/app/`. 

### JavaScript Polyfill

This can currently be mocked out using:

```ts
if (!('host' in window.location)) {
  const url = ((global.nativeModuleProxy || {}).SourceCode || {}).scriptURL;
  const { parse } = require('url');
  window.location = parse(url);
}
```

## `Location.reload()`

> [API reference](https://developer.mozilla.org/en-US/docs/Web/API/Location/reload)

Another highly important feature is the ability to reload the JS bundle, this is useful for hot module reloading in development, or testing persistent modules like k/v storage.

### JavaScript Polyfill

This can currently be mocked out using:

> Notice this currently doesn't fully work in the Expo Go runtime due to caching issues.

```ts
if (!('reload' in window.location)) {
  window.location.reload = () => {
    const { DevSettings } = require('react-native');
    DevSettings.reload();
  };
}
```

## `Location.assign()`

> [API reference](https://developer.mozilla.org/en-US/docs/Web/API/Location/assign)

Assign can be used to completely swap the existing JS bundle for another at a given resource path. This could be an interesting addition to navigation libraries as they could fully unload the resources in the engine.

### JavaScript Polyfill

It's unclear how to currently perform this operation from JavaScript using existing native functionality exposed by the React Native runtime.

## `Location.replace()`

> [API reference](https://developer.mozilla.org/en-US/docs/Web/API/Location/replace)

Since we currently do not implement the [History API](https://developer.mozilla.org/en-US/docs/Web/API/History) this method would be identical to `Location.assign()` as nothing is added to the History API.

## `Location.toString()`

> [API reference](https://developer.mozilla.org/en-US/docs/Web/API/Location/toString)

Since we a utility function for retrieving the stringified results, simply returns the `Location.href`.

### JavaScript Polyfill

This can currently be mocked out using:

> Notice this depends on the properties being implemented as described.

```ts
if (!('toString' in window.location)) {
  window.location.toString = function() {
    return this.href;
  };
}
```

# Hosting Services

In `expo-updates` we may want to support having the production version of `location.href` refer to the `expo-updates` hosted location (`https://u.expo.dev/user/slug/`), which `expo-updates` would then be responsible for intercepting and swapping network requests for cached resources based on a caching policy (i.e. a service worker).

# Implementation

I propose we create a package like `@expo-runtime/location` (naming RFC), which is installed in the `expo` package and required to be inside every Expo project. We'd start depending on it in our bundler tooling in the future, which would fallback on a JS polyfill which console warns instead of attempting to use random internal native functionality. 

This package would have no public JS API, and would entirely be used with Expo autolinking. The API would be injected into the global JS context via native code on iOS and Android code. Web obviously wouldn't need any kinda polyfill as the browser implements this feature.

**Comment by ide** [comment]
Directionally this seems right for Expo to be one of the platforms like the web and Deno that implement part of the industry standard HTML specification (important note: I don't mean HTML the language and the DOM).

I don't want to create another npm org and would rather use `@expo`. Some possible naming conventions are `@expo/standard-web-location` and `@expo/standard-location`. We'd probably want to rename `expo-standard-web-crypto` to match.

It's not mentioned in this RFC but I lean towards not globally defining location automatically. That is the programmer's choice. They can run:
```
import { location } from `@expo/standard-location`;
global.location = Location;

// run any other code that assumes location is globally defined
```
This way we get the best of both worlds.

I suspect the implementation will depend on expo-updates, the package that manages the "brain" (the update) of every Expo app. The modern APIs to access the manifest and information about where the update was loaded from and to reload the update are (or likely should be) part of expo-updates.

**Comment by shirakaba** [comment]
Just a quick thought about whether efforts here could be reused for other Expo runtimes (like the Flutter adapter and the NativeScript adapter).

> This package would have no public JS API, and would entirely be used with Expo autolinking. The API would be injected into the global JS context via native code on iOS and Android code.

Even if there is no public JS API, would the intention still be to organise this as an Expo Unimodule? Or does the JSI usage you mentioned unavoidably couple it to React Native?

Or is this effectively so thin that there’d be nothing to install beyond the native dependency (e.g. the CocoaPod) and thus Expo Unimodules wouldn’t even come into it?

**Reply by EvanBacon** [reply]
> Even if there is no public JS API, would the intention still be to organise this as an Expo Unimodule? Or does the JSI usage you mentioned unavoidably couple it to React Native?

We're still deciding how we want to implement the interface. It could be as simple as a turbo module which provides the JSContext, then we simply install the interface directly onto the global, or it could be a bit more true to turbo modules (which would enable lazy loading), we would have runtime code which could be used as a ponyfill/polyfill like @ide [mentioned](https://github.com/expo/expo/discussions/14385#discussioncomment-1309177).

We could also potentially write part of the interface in JavaScript, and embed that JS inside of the native module as a "builtin", this system is used in [WebKit](https://github.com/WebKit/WebKit/blob/main/Source/JavaScriptCore/builtins/ArrayIteratorPrototype.js), and [Node.js](https://github.com/nodejs/node/blob/master/lib/path.js). The main benefit being that you wouldn't have a decoupled interface where the JS version mismatches the native version (an existing issue that React Native users often face).

> Just a quick thought about whether efforts here could be reused for other Expo runtimes (like the Flutter adapter and the NativeScript adapter).

Our priority will be to ensure that this system is optimal for Expo (and by extension React Native) projects, and to keep it as thin and simple as we can (global -> add function, constant, host object).
If we have the option to make it work better in Flutter, NativeScript etc, or keep it simple and easy to add contributions to, then we'll probably opt towards simplicity.
